pkg/controller/certificates/approver/groupapprove.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package approver
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	authorizationapi "k8s.io/kubernetes/pkg/apis/authorization/v1beta1"
+	capi "k8s.io/kubernetes/pkg/apis/certificates/v1beta1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	certificatesinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/externalversions/certificates/v1beta1"
+	"k8s.io/kubernetes/pkg/controller/certificates"
+)
+
+const (
+	selfNodeClientCertAccess = "certificatessigningrequests/selfnodeclientcert"
+	nodeClientCertAccess     = "certificatessigningrequests/nodeclientcert"
+	nodeServerCertAccess     = "certificatessigningrequests/nodeservercert"
+)
+
+func NewCSRApprovingController(client clientset.Interface, csrInformer certificatesinformers.CertificateSigningRequestInformer) (*certificates.CertificateController, error) {
+	approver := &groupApprover{
+		client: client,
+	}
+	return certificates.NewCertificateController(
+		client,
+		csrInformer,
+		approver.handle,
+	)
+}
+
+// groupApprover implements AutoApprover for signing Kubelet certificates.
+type groupApprover struct {
+	client clientset.Interface
+}
+
+func (ga *groupApprover) handle(csr *capi.CertificateSigningRequest) error {
+	// short-circuit if we're already approved or denied
+	if approved, denied := certificates.GetCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+	csr, err := ga.autoApprove(csr)
+	if err != nil {
+		return fmt.Errorf("error auto approving csr: %v", err)
+	}
+	_, err = ga.client.Certificates().CertificateSigningRequests().UpdateApproval(csr)
+	if err != nil {
+		return fmt.Errorf("error updating approval for csr: %v", err)
+	}
+	return nil
+}
+
+func (cc *groupApprover) autoApprove(csr *capi.CertificateSigningRequest) (*capi.CertificateSigningRequest, error) {
+	x509cr, err := capi.ParseCSR(csr)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("unable to parse csr %q: %v", csr.Name, err))
+		return csr, nil
+	}
```

Comment:
I expected this to be structured more in terms of recognizers and associated permission required.

```go
type CSRRecognizer interface {
  Matches(csrObj, parsedCSR) bool
}

recognizers := []struct{
  recognizer CSRRecognizer
  permission authorizer.ResourceAttributes
}{
  {
    recognizer: isSelfNodeClientCert,
    permission: authorizer.ResourceAttributes{Group: "certificates.k8s.io", Resource: "certificatesigningrequests", Subresource: "selfnodeclient"},
  },
  {
    recognizer: isNodeClientCert,
    permission: authorizer.ResourceAttributes{Group: "certificates.k8s.io", Resource: "certificatesigningrequests", Subresource: "nodeclient"},
  },
  {
    recognizer: isSelfNodeClientCert,
    permission: authorizer.ResourceAttributes{Group: "certificates.k8s.io", Resource: "certificatesigningrequests", Subresource: "selfnodeclient"},
  },
  {
    recognizer: isNodeClientCert,
    permission: authorizer.ResourceAttributes{Group: "certificates.k8s.io", Resource: "certificatesigningrequests", Subresource: "nodeclient"},
  },
}

func autoApprove(csrObj) {
  // short-circuit on pre-approved
  parsedCSR, err := parse(csrObj)
  // short-circuit on error
  for _, r := range recognizers {
    if !r.recognizer.Matches(csrObj, parsedCSR) {
      continue
    }
    approved, err := authorize(csrObj, r.permission)
    if err != nil {
      // accumulate error in case we ultimately fail?
      continue
    }
    if approved {
      // record approval
    }
  }
}
```